contributors/design-proposals/watchable-discovery.md

```diff
@@ -0,0 +1,231 @@
+allowed verbs.
+
+The default RBAC rules will make the resources readable for every
+authenticated user.
+
+It would be possible to add a flag for unauthenticated readability,
+i.e. as a counterpart to the `kube-public` API group that is used
+during kubelet bootstrapping. But this is a non-goal for now.
+
+The discovery information includes its own API group.
+
+## New API Types
+
+The proposed API group name is `discovery.k8s.io`, starting with a
+`v1alpha1` version.
+
+The only new Kind is `Group`, containing all version information and
+all resources. The name of group objects is the API group name. The
+Kind is cluster-scoped.
+
+```golang
+// +genclient=true
+// +nonNamespaced=true
+
+// Group describes an API group with all its versions and resources.
+type Group struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Status of the group.
+	Status GroupStatus `json:"spec,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+
+// GroupStatus describes the API group with all its versions and resources.
+type GroupStatus struct {
+	// versions are the versions supported in this group, order by preference. The first one is the preferred
+	// version for the group.
+	Versions []GroupVersion `json:"versions" protobuf:"bytes,1,rep,name=versions"`
+	// a map of client CIDR to server address that is serving this group.
+	// This is to help clients reach servers in the most network-efficient way possible.
+	// Clients can use the appropriate server address as per the CIDR that they match.
+	// In case of multiple matches, clients should use the longest matching CIDR.
+	// The server returns only those CIDRs that it thinks that the client can match.
+	// For example: the master will return an internal IP CIDR only, if the client reaches the server using an internal IP.
+	// Server looks at X-Forwarded-For header or X-Real-Ip header or request.RemoteAddr (in that order) to get the client IP.
+	ServerAddressByClientCIDRs []metav1.ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs" protobuf:"bytes,2,rep,name=serverAddressByClientCIDRs"`
+}
+
+// GroupVersion describes one version of an API group, including all resources available in this version.
+type GroupVersion struct {
+	// the version name
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// resources contains the name of the resources and if they are namespaced.
+	Resources []metav1.APIResource `json:"resources" protobuf:"bytes,2,rep,name=resources"`
+}
+
+// GroupList is a list of API groups.
+type GroupList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// List of group.
+	Items []Group `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+```
+
+## Behavior
+
+Each API server in the cluster serves its own `discovery.k8s.io` API
```

Comment:
> I don't agree with each apiserver having a private copy of this, nothing else in the system is done this way?
> 
> I don't think apiserver should be doing anything here, actually. I tentatively agree with the discovery types and group, but there should just be a few controllers syncing them up with what is currently stored in TPR & and apiregistration groups.

Each API server needs to service discovery information and ought to be able to have a watchable API that describes it.  In addition, to avoid some of the coupling we currently have in our API installer path, driving the actual `/apis` endpoints from an API like this will reduce internal coupling.

In addition, by having the constituent API servers provide `discovery.k8s.io`, the aggregated API server would be able to watch them to drive a unified view easily an cleanly via composition.